dag_in_context/src/pretty_print.rs

```diff
@@ -0,0 +1,753 @@
+                let right = self.refactor_shared_expr(y, fold_when, to_rust, log);
+                Expr::Concat(Rc::new(left), Rc::new(right))
+            }
+            Expr::Switch(x, inputs, _branches) => {
+                let cond = self.refactor_shared_expr(x, fold_when, to_rust, log);
+                let inputs = self.refactor_shared_expr(inputs, fold_when, to_rust, log);
+                let branches = _branches
+                    .iter()
+                    .map(|branch| {
+                        Rc::new(self.refactor_shared_expr(branch, fold_when, to_rust, log))
+                    })
+                    .collect::<Vec<_>>();
+                let switch = Expr::Switch(Rc::new(cond), Rc::new(inputs), branches);
+                fold_or_plain(self, switch, log)
+            }
+            Expr::If(x, inputs, y, z) => {
+                let pred = self.refactor_shared_expr(x, fold_when, to_rust, log);
+                let inputs = self.refactor_shared_expr(inputs, fold_when, to_rust, log);
+                let left = self.refactor_shared_expr(y, fold_when, to_rust, log);
+                let right = self.refactor_shared_expr(z, fold_when, to_rust, log);
+                let if_expr = Expr::If(
+                    Rc::new(pred),
+                    Rc::new(inputs),
+                    Rc::new(left),
+                    Rc::new(right),
+                );
+                fold_or_plain(self, if_expr, log)
+            }
+            Expr::DoWhile(inputs, body) => {
+                let inputs = self.refactor_shared_expr(inputs, fold_when, to_rust, log);
+                let body = self.refactor_shared_expr(body, fold_when, to_rust, log);
+                let dowhile = Expr::DoWhile(Rc::new(inputs), Rc::new(body));
+                fold_or_plain(self, dowhile, log)
+            }
+            Expr::Arg(ty, assum) => {
+                let ty = self.refactor_shared_type(ty, log);
+                let assum_str = self.refactor_shared_assum(assum, fold_when, to_rust, log);
+                Expr::Arg(ty, Assumption::WildCard(assum_str))
+            }
+            Expr::Symbolic(_) => panic!("No symbolic should occur here"),
+        }
+    }
+}
+
+impl Expr {
+    pub fn pretty(&self) -> String {
+        let (term, termdag) = Rc::new(self.clone()).to_egglog();
+        let expr = termdag.term_to_expr(&term);
+        expr.to_sexp().pretty()
+    }
+
+    fn gather_concat_children(&self) -> Vec<String> {
+        match self {
+            Expr::Concat(lhs, rhs) => {
+                let mut lhs = lhs.as_ref().gather_concat_children();
+                let mut rhs = rhs.as_ref().gather_concat_children();
+                lhs.append(&mut rhs);
+                lhs
+            }
+            Expr::Single(expr) => {
+                let expr = Self::to_ast(expr.as_ref());
+                vec![expr]
+            }
+            _ => vec![self.to_ast()],
+        }
+    }
+
+    pub fn abbrev(&self) -> String {
```

Comment:
this function will either go away or you can incorporate it to use fresh numbers

Reply:
I would like make fresh seperate by taking an info.